ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Block/SupportsDiffusionPipelining.hs

```diff
@@ -0,0 +1,273 @@
+  , DisableDiffusionPipelining (..)
+    -- ** 'SelectViewDiffusionPipelining'
+  , SelectViewDiffusionPipelining (..)
+  , SelectViewTentativeState (..)
+    -- * Data family instances
+  , BlockConfig (..)
+  , Header (..)
+  ) where
+
+import           Control.Monad (guard)
+import           Data.Coerce
+import           Data.Kind
+import           Data.Proxy
+import           GHC.Generics (Generic)
+import           NoThunks.Class
+import           Ouroboros.Consensus.Block.Abstract
+import           Ouroboros.Consensus.Block.SupportsProtocol
+import           Ouroboros.Consensus.Protocol.Abstract
+
+-- | Block functionality required to support __Block Diffusion Pipelining via
+-- Delayed Validation__ (DPvDV).
+--
+-- == High-level context
+--
+-- With DPvDV, a node is, under certain conditions, already announcing a new
+-- block to its downstream peers /before/ it has fully validated the block body.
+-- Concretely, the node maintains a /tentative header/ that, if present, extends
+-- the current selection, and is announced via the ChainSync servers to
+-- downstream peers.
+--
+-- Ideally, the body turns out to be valid, in which case the tentative header
+-- is cleared, and the selection is updated to now include the header at its
+-- tip.
+--
+-- If the body corresponding to the tentative header turns out to be invalid (we
+-- call such a header a /trap header/), the tentative header is cleared, and the
+-- ChainSync servers send a rollback instruction. In this case, the network
+-- wasted work in diffusing, storing and validating this block. If abused, this
+-- could cause an unbounded amount of work for honest nodes. Hence, we need to
+-- enforce that our upstream nodes adhere to an appropriate criterion related to
+-- trap headers, and so must also restrict ourselves accordingly such that
+-- downstream nodes do not disconnect from us.
+--
+-- == Semantics
+--
+-- This type class allows to define a block-specific criterion determining
+-- whether a header that might turn out to be a trap header is allowed to be set
+-- as the tentative header.
+--
+-- This is used in two places:
+--
+--  - The ChainSel logic. We make sure that we only set the tentative header if
+--    this criterion is fulfilled.
+--
+--  - The BlockFetch clients, in combination with the ChainSel validation logic.
+--    For every upstream BlockFetch peer, we make sure that the invalid blocks
+--    this peer sent adhere to the pipelining criterion.
+--
+-- Concretely, the abstract Consensus layer maintains 'TentativeHeaderState's
+-- (one in ChainSel, and one for each (BlockFetch) upstream peer). Suppose that
+-- @hdr@ either might turn out or is already known to be a trap header. Then
+--
+-- @'applyTentativeHeaderView' ('Proxy' \@blk) thv st@
+--
+-- (where @thv = 'tentativeHeaderView' bcfg hdr@) will return
+--
+--  - 'Nothing' if @hdr@ does not satisfy the pipelining criterion.
+--
+--      - In ChainSel, this means that @hdr@ should not be pipelined, as it might
```

Comment:
> as it might turn out to be a trap header

I suggest either removing that clause, or changing it to

> as it would violate the criterion if it turns out to be a trap header